gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/vertx/web/handler/impl/internal/mfa/MFAChallengeStep.java

```diff
@@ -45,61 +45,66 @@ public MFAChallengeStep(Handler<RoutingContext> wrapper, RuleEngine ruleEngine,
-                challenge(routingContext, flow);
+                challenge(context, flow);
             } else if (isChallengeActive(client)) {
                 switch (getChallengeSettings(client).getType()) {
-                    case REQUIRED -> required(routingContext, flow, context);
-                    case CONDITIONAL -> conditional(routingContext, flow, client, context);
-                    case RISK_BASED -> riskBased(routingContext, flow, client, context);
+                    case REQUIRED -> required(context, flow);
+                    case CONDITIONAL -> conditional(context, flow, client);
+                    case RISK_BASED -> riskBased(context, flow, client);
                 }
             } else {
-                continueFlow(routingContext, flow);
+                continueFlow(context, flow);
             }
         } else {
-            continueFlow(routingContext, flow);
+            continueFlow(context, flow);
         }
     }
 
-    private void required(RoutingContext routingContext, AuthenticationFlowChain flow, MfaFilterContext context) {
-        if (context.isChallengeOnceCompleted() && isRememberDeviceOrSkipped(context)) {
-            continueFlow(routingContext, flow);
+    private void required(MfaFilterContext context, AuthenticationFlowChain flow) {
+        if (context.isChallengeOnceCompleted() || isRememberDevice(context)) {
+            continueFlow(context, flow);
         } else {
-            challenge(routingContext, flow);
+            challenge(context, flow);
         }
     }
 
-    private void conditional(RoutingContext routingContext, AuthenticationFlowChain flow, Client client, MfaFilterContext context) {
-        if (context.isChallengeOnceCompleted() || challengeConditionSatisfied(client, context, ruleEngine)) {
-            continueFlow(routingContext, flow);
+    private void conditional(MfaFilterContext context, AuthenticationFlowChain flow, Client client) {
+        if (challengeConditionSatisfied(client, context, ruleEngine)) {
+            continueFlow(context, flow);
+        } else if (context.getRememberDeviceSettings().isSkipChallengeWhenRememberDevice()) {
+            required(context, flow);
+        } else if (context.isChallengeOnceCompleted()) {
+            continueFlow(context, flow);
         } else {
-            challenge(routingContext, flow);
+            challenge(context, flow);
         }
     }
 
-    private void riskBased(RoutingContext routingContext, AuthenticationFlowChain flow, Client client, MfaFilterContext context) {
-        if (context.isChallengeOnceCompleted() || isSafe(client, context)) {
-            continueFlow(routingContext, flow);
+    private void riskBased(MfaFilterContext context, AuthenticationFlowChain flow, Client client) {
+        if (isSafe(context, client)) {
+            continueFlow(context, flow);
+        } else if (context.isChallengeOnceCompleted()) {
+            continueFlow(context, flow);
```

Comment:
**question**: is there any reason why you move it in 2 differents statement instead of using a or opertor ?